src/components/AceEditor.jsx

```diff
@@ -0,0 +1,176 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import bindAll from 'lodash-es/bindAll';
+import constant from 'lodash-es/constant';
+import get from 'lodash-es/get';
+import throttle from 'lodash-es/throttle';
+import noop from 'lodash-es/noop';
+
+import {EditorLocation} from '../records';
+import {createAceEditor, createAceSessionWithoutWorker} from '../util/ace';
+
+import 'brace/ext/searchbox';
+import 'brace/mode/html';
+import 'brace/mode/css';
+import 'brace/mode/javascript';
+import 'brace/theme/monokai';
+
+const RESIZE_THROTTLE = 250;
+const NORMAL_FONTSIZE = 14;
+const LARGE_FONTSIZE = 20;
+
+class AceEditor extends React.Component {
+  constructor() {
+    super();
+
+    this._handleWindowResize = throttle(() => {
+      if (this._editor) {
+        this._resizeEditor();
+      }
+    }, RESIZE_THROTTLE);
+
+    bindAll(
+      this,
+      '_handleWindowResize',
+      '_resizeEditor',
+      '_setupEditor',
+    );
+
+    this.render = constant(
+      <div
+        className="editors__editor"
+        ref={this._setupEditor}
+      />,
+    );
+  }
+
+  componentDidMount() {
+    this._focusRequestedLine(this.props.requestedFocusedLine);
+    this._applyFontSize(this.props.textSizeIsLarge);
+    window.addEventListener('resize', this._handleWindowResize);
+  }
+
+  componentDidUpdate({
+    percentageOfHeight: prevPercentageOfHeight,
+    projectKey: prevProjectKey,
+    source: prevSource,
+  }) {
+    const {
+      errors,
+      percentageOfHeight,
+      projectKey,
+      requestedFocusedLine,
+      source,
+      textSizeIsLarge,
+    } = this.props;
+
+    if (projectKey !== prevProjectKey) {
+      this._startNewSession(source);
+    } else if (source !== prevSource && source !== this._editor.getValue()) {
+      this._editor.setValue(source, 1);
+    }
+
+    this._focusRequestedLine(requestedFocusedLine);
+    this._applyFontSize(textSizeIsLarge);
+
+    if (percentageOfHeight !== prevPercentageOfHeight) {
+      requestAnimationFrame(this._resizeEditor);
+    }
+
+    this._editor.getSession().setAnnotations(errors);
+  }
+
+  componentWillUnmount() {
+    this._editor.destroy();
+    window.removeEventListener('resize', this._handleWindowResize);
+  }
+
+  _focusRequestedLine(requestedFocusedLine) {
+    if (get(requestedFocusedLine, 'component') !==
+      `editor.${this.props.language}`) {
+      return;
+    }
+
+    this._editor.moveCursorTo(
+      requestedFocusedLine.line,
+      requestedFocusedLine.column,
+    );
+
+    this._scrollToLine(requestedFocusedLine.line);
+    this._editor.focus();
+    this.props.onRequestedLineFocused();
+  }
+
+  _resizeEditor() {
+    this._editor.resize();
+  }
+
+  _scrollToLine(lineNumber) {
+    const shouldCenterVertically = true;
+    const shouldAnimate = true;
```

Comment:
If these two consts will always be true, why not just directly set both to `true` as arguments? If it's supposed to be configurable in the future, these values should be located in a properties/config file.

Reply:
Not sure/good point but this is basically just moving the old `Editor` component into a new place, and we’re deprecating it anyway, so I don’t think there’s any need to worry about this question